docs/design-documents/20240430-schema-support.md

```diff
@@ -0,0 +1,420 @@
+IP address and the gRPC port. It's safe to assume that in most, if not all, real
+world use cases, Conduit and connectors will be running on the same host, so we
+can assume that the host is `localhost`. The gRPC port can be communicated to
+the connector via an environment variable.
+
+This service is intended to be used by connectors only. To facilitate that,
+Conduit can generate tokens that connectors would use to authenticate with
+Conduit.
+
+The service should run on a random port. In VMs with hardened security, that
+might not be always possible, so it should be possible for the schema service to
+run on a pre-defined port.
+
+A skeleton of the gRPC definition of the service would be:
+
+```protobuf
+syntax = "proto3";
+
+service SchemaService {
+  rpc Create(CreateSchemaRequest) returns (CreateSchemaResponse);
+  rpc Fetch(FetchSchemaRequest) returns (FetchSchemaResponse);
+}
+
+message CreateSchemaRequest {
+  string name = 1;
+  // other fields
+}
+
+message CreateSchemaResponse {
+  string id = 1;
+}
+
+message FetchSchemaRequest {
+  string id = 1;
+}
+message FetchSchemaResponse {}
+```
+
+**Advantages**:
+
+1. Easy to understand: the gRPC methods, together with requests and responses,
+   can easily be understood from a proto file.
+2. An HTTP API for the schema registry can easily be exposed (if needed).
+3. This API can be extended to include other methods that a connector might
+   need (e.g. connector storage)
+
+**Disadvantages**:
+
+1. Changes needed to communicate Conduit's gRPC port to the connector.
+2. Streams are faster that gRPC method calls. However, registering or fetching a
+   schema is an infrequent operation, so this is not a concern for us.
+
+#### Chosen option
+
+**Option 2** is the chosen method since it offers more clarity and the support
+for remote Conduit instances.
+
+## Required changes
+
+### Conduit
+
+Conduit needs to expose a gRPC schema service as explained above. The gRPC
+service exposes methods needed for connectors to work with schemas. Initially,
+the service will use the Apicurio Registry to actually manage the schemas. Later
+on, we will migrate to our own schema registry.
+
+The service's port will be random and Conduit will make it available to
+connectors via an environment variable.
+
+### Conduit Commons
+
+Conduit Commons needs to provide the following functions that will be used by
+multiple libraries (Connector SDK, Processor SDK):
+
+1. A function that creates an Avro schema
+2. A function that encodes values using an Avro schema
+3. A function decodes a slice of bytes into a value, using an Avro schema
+
+### Connector SDK
+
+The Connector SDK needs to provide the following functions:
+
+1. A function that registers a schema.
+2. A function that fetches a schema.
+
+### Processor SDK
+
+The Processor SDK needs to provide the following functions:
+
+1. A function that registers a schema.
+2. A function that fetches a schema.
```

Comment:
This should go into conduit commons.

Reply:
The reason I put it here is because we might need different mechanisms for connectors and processors. For standalone connectors it makes sense to gRPC calls to Conduit. For built-in ones, maybe less so. But standalone processors won't be able to do anything on the network, so we'll probably need to do it through the usual commands-and-response thingy we have. I believe we can figure out the details once we come to that.

Reply:
Oh, you're right, these can't go into commons, my bad!

Reply:
That's fine, it's always good to double check.:)